internal/iam/authorizers.go

```diff
@@ -0,0 +1,210 @@
+	}
+	return OAuthGrantTypeServicePrincipal
+}
+
+// GetResourceManagementAuthorizer gets an OAuthTokenAuthorizer for Azure Resource Manager
+func GetResourceManagementAuthorizer() (autorest.Authorizer, error) {
+	if armAuthorizer != nil {
+		return armAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(
+		grantType(), config.Environment().ResourceManagerEndpoint)
+
+	if err == nil {
+		// cache
+		armAuthorizer = a
+	} else {
+		// clear cache
+		armAuthorizer = nil
+	}
+	return armAuthorizer, err
+}
+
+// GetBatchAuthorizer gets an OAuthTokenAuthorizer for Azure Batch.
+func GetBatchAuthorizer() (autorest.Authorizer, error) {
+	if batchAuthorizer != nil {
+		return batchAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(
+		grantType(), config.Environment().BatchManagementEndpoint)
+
+	if err == nil {
+		// cache
+		batchAuthorizer = a
+	} else {
+		// clear cache
+		batchAuthorizer = nil
+	}
+
+	return batchAuthorizer, err
+}
+
+// GetGraphAuthorizer gets an OAuthTokenAuthorizer for graphrbac API.
+func GetGraphAuthorizer() (autorest.Authorizer, error) {
+	if graphAuthorizer != nil {
+		return graphAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(grantType(), config.Environment().GraphEndpoint)
+
+	if err == nil {
+		// cache
+		graphAuthorizer = a
+	} else {
+		graphAuthorizer = nil
+	}
+
+	return graphAuthorizer, err
+}
+
+// GetKeyvaultAuthorizer gets an OAuthTokenAuthorizer for use with Key Vault
+// keys and secrets. Note that Key Vault *Vaults* are managed by Azure Resource
+// Manager.
+func GetKeyvaultAuthorizer() (autorest.Authorizer, error) {
+	if keyvaultAuthorizer != nil {
+		return keyvaultAuthorizer, nil
+	}
+
+	// BUG: default value for KeyVaultEndpoint is wrong
+	vaultEndpoint := strings.TrimSuffix(config.Environment().KeyVaultEndpoint, "/")
+	// BUG: alternateEndpoint replaces other endpoints in the configs below
+	alternateEndpoint, _ := url.Parse(
+		"https://login.windows.net/" + config.TenantID() + "/oauth2/token")
+
+	var a autorest.Authorizer
+	var err error
+
+	switch grantType() {
+	case OAuthGrantTypeServicePrincipal:
+		oauthconfig, err := adal.NewOAuthConfig(
+			config.Environment().ActiveDirectoryEndpoint, config.TenantID())
+		if err != nil {
+			return a, err
+		}
+		oauthconfig.AuthorizeEndpoint = *alternateEndpoint
+
+		token, err := adal.NewServicePrincipalToken(
+			*oauthconfig, config.ClientID(), config.ClientSecret(), vaultEndpoint)
+		if err != nil {
+			return a, err
+		}
+
+		a = autorest.NewBearerAuthorizer(token)
+
+	case OAuthGrantTypeDeviceFlow:
+		deviceConfig := auth.NewDeviceFlowConfig(config.ClientID(), config.TenantID())
+		deviceConfig.Resource = vaultEndpoint
+		deviceConfig.AADEndpoint = alternateEndpoint.String()
+		a, err = deviceConfig.Authorizer()
+	default:
+		log.Fatalln("invalid token type specified")
```

Comment:
I'm never a fan of `os.Exit(1)`, like is used by `log.Fatalln`. It prevents any deferred actions from executing all the way up the stack.

I would rather see this clause as:

``` Go
return nil, errors.New("invalid token type specified")
```

or even better, create a global variable `var ErrInvalidTokenType = errors.New("invalid token type specified")`  then return that. :)